libraries/rodeos/include/b1/rodeos/callbacks/chaindb.hpp

```diff
@@ -0,0 +1,882 @@
+            }
+            else
+            {
+               stream = {view_it->get_kv()->value.data(), view_it->get_kv()->value.size()};
+            }
+            auto record = std::get<0>(eosio::from_bin<Ship_index_type>(stream));
+            primary = record.primary_key;
+            it.next = get_iterator(it.table_index, std::move(record), std::move(*view_it));
+            return it.next;
+         }
+      }  // next_impl
+
+      template <typename Ship_index_type, typename F1, typename F2>
+      int32_t prev_impl(int itr,
+                        uint64_t& primary,
+                        bool view_it_is_secondary,
+                        F1 create_view_it,
+                        F2 get_iterator)
+      {
+         if (itr == -1)
+            throw std::runtime_error("decrement invalid iterator");
+         iterator* it = nullptr;
+         if (std::numeric_limits<int32_t>::min() < itr && itr < -1)
+         {
+            size_t table_index = end_iterator_to_index(itr);
+            if (table_index >= end_iterators.size())
+               throw std::runtime_error("decrement non-existing iterator");
+            it = &end_iterators[table_index];
+         }
+         else
+         {
+            if (size_t(itr) >= iterators.size())
+               throw std::runtime_error("decrement non-existing iterator");
+            it = &iterators[itr];
+         }
+         if (it->prev)
+         {
+            if (*it->prev >= 0)
+               primary = iterators[*it->prev].primary;
+            return *it->prev;
+         }
+         std::optional<chain_kv::view::iterator> view_it = std::move(it->view_it);
+         it->view_it.reset();
+         if (!view_it)
+            view_it = create_view_it(*it, itr >= 0);
+         --*view_it;
+         if (view_it->is_end())
+         {
+            it->prev = -1;
+            return *it->prev;
+         }
+         else
+         {
+            eosio::input_stream stream;
+            std::shared_ptr<const chain_kv::bytes> v;
+            if (view_it_is_secondary)
+            {
+               v = view.get(state_account.value, view_it->get_kv()->value);
+               stream = {v->data(), v->size()};
+            }
+            else
+            {
+               stream = {view_it->get_kv()->value.data(), view_it->get_kv()->value.size()};
+            }
+            auto record = std::get<0>(eosio::from_bin<Ship_index_type>(stream));
+            primary = record.primary_key;
+            it->prev = get_iterator(it->table_index, record, std::move(*view_it));
+            return *it->prev;
+         }
+      }  // prev_impl
+
+      // Precondition: std::numeric_limits<int32_t>::min() < ei < -1
+      // Iterator of -1 is reserved for invalid iterators (i.e. when the appropriate table has not
+      // yet been created).
+      size_t end_iterator_to_index(int32_t ei) const { return (-ei - 2); }
+      // Precondition: indx < tables.size() <= std::numeric_limits<int32_t>::max()
```

Comment:
Are you sure about this precondition?  `numeric_limits<int32_t>::max()` will give `-(0x7FFFFFFF + 2) = 0x7FFFFFFE` which is positive and probably not correct.

Reply:
It's not. apply_context.hpp has the same mistake. Adding a note.

Reply:
Which means the "too many open tables" check needs fixing

Reply:
I spoke too soon. If `tables.size() == 0x7fff'ffff`, then index is capped at `0x7fff'fffe`. Since the type is `size_t`:

```
-(0x7fff'fffe + 2) =
-0x8000'0000 =
0xffff'ffff'8000'0000
```

truncate to `int32_t`: `0x8000'0000`.

It's still negative, but its return value doesn't meet the precondition for `end_iterator_to_index`, which also comes from apply_context.hpp.

Reply:
line 85 caps the table *index* at `0x7fffffff`, not the table size.  Note how `table.push_back` comes after the size check.

Reply:
Fixing line 85 to meet the precondition

Reply:
Also tightening the precondition of `index_to_end_iterator` so it doesn't produce `0x8000'0000`, which violates the precondition to `end_iterator_to_index`.